internal/resolution/resolve_vulns.go

```diff
@@ -0,0 +1,149 @@
+package resolution
+
+import (
+	"context"
+	"sync"
+
+	"deps.dev/util/resolve"
+	"github.com/google/osv-scanner/internal/utility/vulns"
+	"github.com/google/osv-scanner/pkg/lockfile"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/osv"
+	"golang.org/x/exp/maps"
+)
+
+// computeVulns scans for vulnerabilities in a resolved graph and populates res.Vulns
+func (res *ResolutionResult) computeVulns(ctx context.Context, cl resolve.Client) error {
+	// TODO: local vulnerability db support
+
+	//nolint:contextcheck // TODO: Should Hydrate be accepting a context?
+	nodeVulns, err := queryOSV(res.Graph)
+	if err != nil {
+		return err
+	}
+	// Find all dependency paths to the vulnerable dependencies
+	var vulnerableNodes []resolve.NodeID
+	vulnInfo := make(map[string]models.Vulnerability)
+	for i, vulns := range nodeVulns {
+		if len(vulns) > 0 {
+			vulnerableNodes = append(vulnerableNodes, resolve.NodeID(i))
+		}
+		for _, vuln := range vulns {
+			vulnInfo[vuln.ID] = vuln
+		}
+	}
+
+	nodeChains := computeChains(res.Graph, vulnerableNodes)
+	vulnChains := make(map[string][]DependencyChain)
+	for i, idx := range vulnerableNodes {
+		for _, vuln := range nodeVulns[idx] {
+			vulnChains[vuln.ID] = append(vulnChains[vuln.ID], nodeChains[i]...)
+		}
+	}
+
+	// construct the ResolutionVulns
+	// TODO: This constructs a single ResolutionVuln per vulnerability ID.
+	// The scan action treats vulns with the same ID but affecting different versions of a package as distinct.
+	// TODO: Combine aliased IDs
+	for id, vuln := range vulnInfo {
+		rv := ResolutionVuln{Vulnerability: vuln, DevOnly: true}
+		for _, chain := range vulnChains[id] {
+			if chainConstrains(ctx, cl, chain, &rv.Vulnerability) {
+				rv.ProblemChains = append(rv.ProblemChains, chain)
+			} else {
+				rv.NonProblemChains = append(rv.NonProblemChains, chain)
+			}
+			rv.DevOnly = rv.DevOnly && ChainIsDev(chain, res.Manifest)
+		}
+		if len(rv.ProblemChains) == 0 {
+			// There has to be at least one problem chain for the vulnerability to appear.
+			// If our heuristic couldn't determine any, treat them all as problematic.
+			rv.ProblemChains = rv.NonProblemChains
+			rv.NonProblemChains = nil
+		}
+		res.Vulns = append(res.Vulns, rv)
+	}
+
+	return nil
+}
+
+// vulnCache caches all vulnerabilities affecting any versions of particular packages.
+// We cache call vulns & manually check affected, rather than querying the affected versions directly
+// since remediation needs to query for OSV vulnerabilities multiple times for the same packages.
+var vulnCache sync.Map // map[resolve.PackageKey][]models.Vulnerability
```

Comment:
is there a way for us to avoid this being a global? this could be surpising for future users of this as a library.

Reply:
I could pass it in as part of the arguments, it needs to be passed in all of `ComputeRelaxPatches` -> `tryRelaxRemediate` -> `Resolve` -> `computeVulns`.

It probably makes sense to bundle an `OSVClient` together with the `resolve.Client` (which would also make supporting local dbs easier) - I might do that in another PR.

Reply:
Ack.  Please  add a TODO here  to  do  that.